File: ui/src/cron_utils.rs
//! Cron-expression parsing and formatting helpers shared across the routines,
//! schedule, and heatmap views. Split out of `main.rs` to keep that file under
//! the line-count gate; re-exported from the crate root so existing
//! `crate::parse_cron`-style paths keep working unchanged.

use croner::Cron;

/// Parse a cron expression into a `Cron`, normalizing the 7-field
/// (sec min hour dom month dow year) form to 5-field to match server behaviour.
/// Returns `None` for empty or invalid expressions.
pub(crate) fn parse_cron(expr: &str) -> Option<Cron> {
    let s = expr.trim();
    if s.is_empty() {
        return None;
    }
    let normalized = if s.starts_with('@') {
        s.to_string()
    } else {
        let parts: Vec<&str> = s.split_whitespace().collect();
        if parts.len() == 7 {
            parts[1..6].join(" ")
        } else {
            s.to_string()
        }
    };
    normalized.parse::<Cron>().ok()
}

/// Returns (is_valid, human description) for a cron expression.
pub(crate) fn describe_cron_live(expr: &str) -> (bool, String) {
    if expr.trim().is_empty() {
        return (false, "— enter a cron expression —".into());
    }
    match parse_cron(expr) {
        Some(cron) => (true, cron.describe()),
        None => (false, "Invalid cron expression".into()),
    }
}

pub(crate) fn reltime(ts: u64) -> String {
    if ts == 0 {
        return "—".into();
    }
    let now = (js_sys::Date::now() / 1000.0) as u64;
    let diff = now.saturating_sub(ts);
    if diff < 60 {
        "just now".into()
    } else if diff < 3_600 {
        format!("{}m ago", diff / 60)
    } else if diff < 86_400 {
        format!("{}h ago", diff / 3_600)
    } else {
        format!("{}d ago", diff / 86_400)
    }
}

File: client/src/lib/cronUtils.ts
import { toString as cronToString } from "cronstrue";

/**
 * Normalizes a 7-field (sec min hour dom month dow year) cron expression to
 * the 5-field form the server understands, matching `croner`'s normalization
 * on the Rust side (`ui/src/cron_utils.rs`). `@`-prefixed shorthand (e.g.
 * `@daily`) and already-5/6-field expressions pass through unchanged.
 */
export function normalizeCron(expr: string): string {
  const s = expr.trim();
  if (s === "" || s.startsWith("@")) return s;
  const parts = s.split(/\s+/);
  return parts.length === 7 ? parts.slice(1, 6).join(" ") : s;
}

/** Returns `[isValid, humanDescription]` for a cron expression. */
export function describeCronLive(expr: string): [boolean, string] {
  if (expr.trim() === "") return [false, "— enter a cron expression —"];
  try {
    return [true, cronToString(normalizeCron(expr), { throwExceptionOnParseError: true })];
  } catch {
    return [false, "Invalid cron expression"];
  }
}

/** Relative-time label for a unix-seconds timestamp: "just now", "5m ago", "3h ago", "2d ago". */
export function reltime(ts: number): string {
  if (ts === 0) return "—";
  const now = Math.floor(Date.now() / 1000);
  const diff = Math.max(0, now - ts);
  if (diff < 60) return "just now";
  if (diff < 3_600) return `${Math.floor(diff / 60)}m ago`;
  if (diff < 86_400) return `${Math.floor(diff / 3_600)}h ago`;
  return `${Math.floor(diff / 86_400)}d ago`;
}

const MONTHS_ABBR = [
  "Jan",
  "Feb",
  "Mar",
  "Apr",
  "May",
  "Jun",
  "Jul",
  "Aug",
  "Sep",
  "Oct",
  "Nov",
  "Dec",
] as const;

/**
 * Absolute local (browser timezone) rendering of a unix-seconds timestamp, e.g.
 * "Jun 21, 2026 12:00". Meant as a tooltip companion to `reltime`'s relative "N ago" text, so
 * hovering reveals wall-clock time. Matches `ui/src/cron_utils.rs`'s `abstime`.
 */
export function abstime(ts: number): string {
  if (ts === 0) return "—";
  const d = new Date(ts * 1000);
  if (Number.isNaN(d.getTime())) return "—";
  const day = String(d.getDate()).padStart(2, "0");
  const hm = `${String(d.getHours()).padStart(2, "0")}:${String(d.getMinutes()).padStart(2, "0")}`;
  return `${MONTHS_ABBR[d.getMonth()]} ${day}, ${d.getFullYear()} ${hm}`;
}
